theseus/optimizer/nonlinear/dogleg.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Any, Dict, Optional, Tuple, Type
+
+import torch
+
+from theseus.core import Objective
+from theseus.optimizer import Linearization
+from theseus.optimizer.linear import LinearSolver
+
+from .trust_region import TrustRegion
+
+
+# See Nocedal and Wright, Numerical Optimization, pp. 73-77
+# https://www.csie.ntu.edu.tw/~r97002/temp/num_optimization.pdf
+class Dogleg(TrustRegion):
+    EPS = 1e-7
+
+    def __init__(
+        self,
+        objective: Objective,
+        linear_solver_cls: Optional[Type[LinearSolver]] = None,
+        vectorize: bool = False,
+        linearization_cls: Optional[Type[Linearization]] = None,
+        linearization_kwargs: Optional[Dict[str, Any]] = None,
+        linear_solver_kwargs: Optional[Dict[str, Any]] = None,
+        abs_err_tolerance: float = 1e-10,
+        rel_err_tolerance: float = 1e-8,
+        max_iterations: int = 20,
+        step_size: float = 1.0,
+        **kwargs,
+    ):
+        super().__init__(
+            objective,
+            linear_solver_cls=linear_solver_cls,
+            vectorize=vectorize,
+            linearization_cls=linearization_cls,
+            linearization_kwargs=linearization_kwargs,
+            linear_solver_kwargs=linear_solver_kwargs,
+            abs_err_tolerance=abs_err_tolerance,
+            rel_err_tolerance=rel_err_tolerance,
+            max_iterations=max_iterations,
+            step_size=step_size,
+            **kwargs,
+        )
+
+    def _compute_delta_impl(self) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        trust_region_2 = self._trust_region**2
+        delta_gn = self.linear_solver.solve()
+        delta_gn_norm_2 = TrustRegion._squared_norm(delta_gn)
+        good_gn_idx = delta_gn_norm_2 < trust_region_2
+        # All Gauss-Newton step are within trust-region, can return
+        if good_gn_idx.all():
+            # Return a False mask since no indices are at the boundary
+            return delta_gn, torch.zeros_like(good_gn_idx).bool()
+
+        # ---------------------------------------------------------------------
+        # Some Gauss-Newton steps are outside trust-region,
+        # need to compute dogleg step
+        # ---------------------------------------------------------------------
+
+        linearization = self.linear_solver.linearization
+
+        neg_grad = linearization.Atb.squeeze(2)
+        Adelta_sd = linearization.Av(neg_grad)
```

Comment:
So this is the computation of the Cauchy step?

Reply:
Yes. Oh, this reminds, me I was going to rename all of these variables. `delta_sd` --> `delta_c` and `t` --> `caucy_step_size` (and also rename the norms accordingly).